### PR TITLE
feat: add noProxy field to repository & repository templates

### LINKED ISSUE #18323: Add `no_proxy` field to repo-creds
# Summary

Add a `no_proxy` field to repo-creds that will be respected when working with a repository matching the repo-cred.

# Motivation

## Use-case 1 - internal helm charts

We @swisspost are using GH in combination with a cluster + helm registry behind a corporate proxy. To reach GH from Argo CD we have created a repo-cred matching our GH org, specifying the proxy to use in the repo-cred. 

Now when we try to add a new Argo CD app that references a GH repository which again has references to helm charts from our internal registry, the proxy is used both for accessing GH but also the private registry (which obviously fails).

## Use-case 2 - kustomize bases from private git

The same enhancement would also help for GH repositories that contain kustomize references to a private git server, which isn't accessible via proxy.

# Workaround

Specifying `NO_PROXY` as env variable globally on the argocd-repo-server solves the problem, but we think a field in the repo creds would be the cleaner solution without any potential side-effects (since the config is only specified where it applies).

# Proposal

Haven't yet looked at the code, but since repo-creds have a field for specifying the proxy to use, I'd not expect it to be that difficult adding no_proxy there as well. In case the enhancement gets accepted I'm willing to contribute this feature.

### PR DESCRIPTION
**Pull request by the-technat**
Adds support for noProxy on repository & repository templates

Closes #18323

Thanks to @chetan-rns for adding the initial first-class proxy support! Adding noProxy to this existing feature was thus made rather easy.

I can confirm that the feature works as intended. It was tested locally and inside Swiss Post's EKS environment (dev stage). If someone wants to test as well, I built the [image](https://github.com/users/the-technat/packages/container/package/argocd) with the PR-commit state.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by LinoSteffen — APPROVED** [review]

**Review by mkilchhofer — APPROVED** [review]
@the-technat gave me a demo in Swiss Posts environment. Worked like a charm 😎

File: docs/operator-manual/declarative-setup.md [hunk 1]
```diff
@@ -484,10 +484,13 @@ stringData:
   type: git
   url: https://github.com/argoproj/private-repo
   proxy: https://proxy-server-url:8888
+  noProxy: ".internal.example.com"
```
Comment (crenshaw-dev) [inline docs/operator-manual/declarative-setup.md:487]: A few places mention that this can be a list. Is it a CSV? If so, could we use a CSV example here?

**Review by crenshaw-dev — DISMISSED** [review]

File: docs/operator-manual/declarative-setup.md (hunk 1 above)
Comment (the-technat) [reply docs/operator-manual/declarative-setup.md]: I updated the example, now it's list and contains different options.

**Review by todaywasawesome — APPROVED** [review]
Looks good!

**Review by agaudreault — APPROVED** [review]

**Review by crenshaw-dev — APPROVED** [review]

**Comment by blakepettersson** [comment]
@the-technat there are some conflicts which needs to be resolved.

**Comment by mkilchhofer** [comment]
❤️ 🙏

**Comment by knweiss** [comment]
IMHO `cmd/util/repo.go` needs this fix:
```go
        command.Flags().StringVar(&opts.Proxy, "proxy", "", "use proxy to access repository")
-       command.Flags().StringVar(&opts.Proxy, "no-proxy", "", "don't access these targets via proxy")
+       command.Flags().StringVar(&opts.NoProxy, "no-proxy", "", "don't access these targets via proxy")
```

**Comment by the-technat** [comment]
Thanks @knweiss for spotting this! Will try to file another PR for this.

**Comment by knweiss** [comment]
I've added issue #21194 for it not to get lost (and give others a chance to find the explanation if they run into problems as we did).

**Pull request merged**